src/Generics/Linear/ViaGHCGenerics.hs

```diff
@@ -0,0 +1,162 @@
+  ) where
+import Data.Coerce (Coercible, coerce)
+import Data.Kind (Constraint, Type)
+import Generics.Linear.Class
+import qualified GHC.Generics as G
+import Unsafe.Coerce
+import GHC.Exts (Any)
+import GHC.TypeLits (TypeError, ErrorMessage (..))
+
+
+-- | When @a@ is an instance
+-- of @"GHC.Generics".'G.Generic'@, @GHCGenerically a@ is an instance
+-- of 'Generic'.
+--
+-- === Warnings
+--
+-- @GHCGenerically@ is intended for use as a 'DerivingVia' target.
+-- Most other uses of its 'Generic' instance will be quite wrong.
+--
+-- @GHCGenerically@ is safe to use with /derived/
+-- @"GHC.Generics".'G.Generic'@ instances, which are linear. If
+-- you choose to use it with a hand-written instance, you should
+-- check that the underlying instance is linear.
+--
+-- === Example
+--
+-- @
+-- data Foo a = Bar a (Either Int a) | Baz (Maybe a) Int
+--   deriving stock (Show, "GHC.Generics".'G.Generic')
+--   deriving 'Generic' via GHCGenerically (Foo a)
+-- @
+newtype GHCGenerically a = GHCGenerically { unGHCGenerically :: a }
+
+instance G.Generic a => Generic (GHCGenerically a) where
+  type Rep (GHCGenerically a) = G.Rep a
+  to = toLinear (GHCGenerically #. G.to)
+  from = toLinear (G.from .# unGHCGenerically)
+
+-- | When @a@ is an instance of @"GHC.Generics".'G.Generic1'@, and its 'G.Rep1'
+-- contains no compositions, @GHCGenerically1 a@ is an instance of 'Generic1'.
+--
+-- === Warning
+--
+-- @GHCGenerically1@ is intended for use as a 'DerivingVia' target.
+-- Most other uses of its 'Generic1' instance will be quite wrong.
+--
+-- @GHCGenerically1@ is safe to use with /derived/
+-- @"GHC.Generics".'G.Generic1'@ instances, which are linear. If
+-- you choose to use it with a hand-written instance, you should
+-- check that the underlying instance is linear.
+--
+-- === Example
+--
+-- @
+-- data Foo a = Bar a (Either Int a) | Baz (Maybe a) Int
+--   deriving stock (Show, "GHC.Generics".'G.Generic1')
+--   deriving 'Generic1' via GHCGenerically1 Foo
+-- @
+type GHCGenerically1 :: forall k. (k -> Type) -> k -> Type
+newtype GHCGenerically1 f a = GHCGenerically1 { unGHCGenerically1 :: f a }
+
+instance (G.Generic1 f, Repairable ('ShowType f) (G.Rep1 f)) => Generic1 (GHCGenerically1 f) where
+  type Rep1 (GHCGenerically1 f) = Repair (G.Rep1 f)
+  to1 :: forall a m. Rep1 (GHCGenerically1 f) a %m-> GHCGenerically1 f a
+  to1 = unsafeCoerce (GHCGenerically1 #. G.to1 @_ @f @a)
+
+  from1 :: forall a m. GHCGenerically1 f a %m-> Rep1 (GHCGenerically1 f) a
+  from1 = unsafeCoerce (G.from1 @_ @f @a .# unGHCGenerically1)
+
+-- | A @"GHC.Generics".'G.Rep1'@ is @Repairable@ if it contains no
+-- compositions. The 'ErrorMessage' argument should be 'ShowType'
+-- of the type whose representation this is.
+type Repairable :: forall k. ErrorMessage -> (k -> Type) -> Constraint
+class Repairable tn grep1 where
+  -- | Convert a @"GHC.Generics".'G.Rep1'@ representation into a 'Rep1'
+  -- representation.
+  type Repair grep1 :: k -> Type
```

Comment:
I'm somewhat surprised that there isn't some kind of `repair :: grep1 x -> Repair grep1 x` method that actually converts a `GHC.Generics`-based `Rep1` instance to a `linear-generics`–based one. My hope would be that you could replace the uses of `unsafeCoerce` above with `repair`, but perhaps there's some complication I'm overlooking.